alg/teca_python_algorithm.py

```diff
@@ -51,58 +55,73 @@ def __getattr__(self, name):
-        """ Override to change number of outputs """
-        return 1
-
     def get_report_callback(self):
         """
-        Returns a function with the signature
-
-            report_callback(port, md_in) -> teca_metadata
-
-        The default implementation passes the report down stream.
-        Override this to customize the behavior of the report
-        phase of execution.
+        returns a callback to be used by the programmable algorithm that
+        forwards calls to the class method.
         """
         def report_callback(port, md_in):
-            import teca_py
-            return teca_py.teca_metadata(md_in[0])
+            return self.report(port, md_in)
         return report_callback
 
     def get_request_callback(self):
         """
-        Returns a function with the signature
-
-            request_callback(port, md_in, req_in) -> [teca_metadata]
-
-        The default implementation passes the request up stream.
-        Override this to customize the behavior of the request
-        phase of execution.
+        returns a callback to be used by the programmable algorithm that
+        forwards calls to the class method.
         """
         def request_callback(port, md_in, req_in):
-            import teca_py
-            return [teca_py.teca_metadata(req_in)]
+            return self.request(port, md_in, req_in)
         return request_callback
 
     def get_execute_callback(self):
         """
-        Returns a function with the signature
-
-            execute_callback(port, data_in, req_in) -> teca_dataset
-
-        The default implementation shallow copies the input dataset.
-        Override this to customize the behavior of the execute
-        phase of execution.
+        returns a callback to be used by the programmable algorithm that
+        forwards calls to the class method.
         """
         def execute_callback(port, data_in, req_in):
-            import teca_py
-            if len(data_in):
-                data_out = data_in[0].new_instance()
-                data_out.shallow_copy(teca_py.as_non_const_teca_dataset(data_out))
-            return data_out
+            return self.execute(port, data_in, req_in)
         return execute_callback
+
+    def get_number_of_input_connections(self):
+        """
+        return the number of input connections this algorithm needs.
+        The default is 1, override to modify.
+        """
+        return 1
+
+    def get_number_of_output_ports(self):
+        """
+        return the number of output ports this algorithm provides.
+        The default is 1, override to modify.
+        """
+        return 1
+
+    def report(self, port, md_in):
+        """
+        return the metadata decribing the data available for consumption.
+        Override this to customize the behavior of the report phase of
+        execution. The default passes metadata on the first input through.
+        """
+        import teca_py
```

Comment:
I'd suggest moving this import to the top of the file, following the PEP 8 style guide for imports: https://www.python.org/dev/peps/pep-0008/#imports